[Feature] Add in-place loading method to QuantumModel

inafergra opened the pull request:
Fixes #484 

This PR proposes a new method in QuantumModel, `load_params_from_dict()` which allows to load the parameters from a dictionary directly to the current model. There is already a similar method `.from_dict()` but that one is a class method that initializes a new instance from the configuration saved in the state dictionary. Adding `load_params_from_dict()` fixes the bug mentioned in #484 as it allows to load the parameters from the dict 'in-place' without initializing a new instance. This is especially important in case the optimizer is defined before loading a checkpoint (as in the qadence's `train()` function) otherwise the optimizer will point to the old model and the parameters of the newly instantiated model won't be updated.

Aside from this, I also reworked the tests so that we properly test loading and saving ckpts. The main contribution here is that now the models are always modified after being saved, so that when they are loaded again the tests won't pass if the loading is not successful (this was not the case before).

Lastly, I also modified the line below so that it also works with torch modules with no `.device()` method. This assumes that all parameters in the module are saved in the same device but I think it's a pretty safe guess
 https://github.com/pasqal-io/qadence/blob/d478f8a7d9c2b3dd1f011099beb2c00af23fe20f/qadence/ml_tools/saveload.py#L75

inafergra commented on qadence/model.py:
```diff
@@ -432,6 +432,30 @@ def _from_dict(cls, d: dict, as_torch: bool = False) -> QuantumModel:
+            d (dict): Dictionary
+        """
+        param_dict = d["param_dict"]
+        for n, param in param_dict.items():
+            if n in param_dict:
```
Indeed thanks, this line can go

chMoussa replied:
Just replace by `if n is in self._params` as you try to access it one line below

inafergra replied:
Actually, wouldn't it be better to let the `try`/`except` statement catch the error in the case where the key does not exist? Otherwise it will go silent and the user might think that all the parameters are loaded correctly when in reality they were not

inafergra replied:
A more elegant solution would be to check what are the `missing_keys` and the `unexpected_keys` in the dictionary, as they do in the [`torch.nn.Module.load_state_dict`]( https://github.com/pytorch/pytorch/blob/d95756f6a5a65be0a39bdfa970fe3f9608a19288/torch/nn/modules/module.py#L2460) method in torch

chMoussa replied:
Indeed, so either we do the check, or simply catch the KeyError. Timeline is up to you.

inafergra replied:
Great, I think it makes sense to do the check. I'll add it here

inafergra commented:
@chMoussa I added the missing/unexpected keys check and an additional argument in the method to enforce that the parameter keys of the dictionary strictly match the param keys in the model. I also added a test for the method in `test_quantum_model.py`

chMoussa commented on qadence/model.py:
```diff
@@ -432,6 +432,56 @@ def _from_dict(cls, d: dict, as_torch: bool = False) -> QuantumModel:
+            else:
+                unexpected_keys.append(n)
+
+        for n in self._params.keys():
```
You can use here `missing_keys = set(self._params.keys()) - set(param_dict.keys())`

chMoussa commented on qadence/model.py:
```diff
@@ -432,6 +432,56 @@ def _from_dict(cls, d: dict, as_torch: bool = False) -> QuantumModel:
+        """
+        param_dict = d["param_dict"]
+        missing_keys = []
+        unexpected_keys = []
```
You can use here `unexpected_keys = set(self._params.keys()) - set(param_dict.keys())` and I would then check if strict and directly raise the error before looping. Loop can then only be done over the present keys. Just a small trick to avoid unnecessary instructions.

inafergra replied:
Wow that's really good! Thanks for that, I'll add it now for both cases and throw the error before

chMoussa reviewed (approved):
Lgtm now, thanks!
